fix handling of sms:, smsto:, mms:, mmsto: URIs

rhodey opened the pull request:
1) added VIEW action and BROWSABLE category to the intent filter of SmsSendtoActivity
2) created two helper methods to make up for bugs in the processing of RFC5724 SMS URIs by android.net.Uri

Fixes #2578
// FREEBIE

tested on my Nexus 5 (A 5.1) & LG Optimus 2X (A 2.3.4).

moxie0 commented:
this is a lot of code, what is the bug you're trying to fix?

rhodey commented:
@moxie0 please see my most recent comment in issue #2578, included in this pull by reference.

moxie0 commented:
why are you parsing this uri by hand instead of using Uri or URI?

rhodey commented:
@moxie0 calling `Uri.getQueryParameter()` on `sms:+437722735932?body=hello%20there` throws an exception stating `this is not a hierarchical url`, none of the Uri query parameter methods work with RFC5724 uris.

Let me try creating a URI from it and using their methods...

rhodey commented:
URI is also unable to cope with RFC5724, throwing `java.net.URISyntaxException: Scheme-specific part expected at index 4: sms:` from its constructor.

I don't believe Uri or URI can be used in this situation unless we do something weird and abuse the `sms:, smsto:, mms:` etc uris to play nicely with the Uri or URI expectations-- which itself would require some parsing by hand.

Would throwing these methods into util and adding unit tests help the cause, @moxie0?

moxie0 commented:
yes you'll have to write a uri class for this if there's nothing available, along with tests for the pathological cases to ensure you don't end up with underflows or overflows

moxie0 commented:
Validate as you parse, and parse everything on construction time and then have getters that return the parsed parts.  Use split instead of indexof and substring (there should be three logical splits, ":", "?", and "&"), don't worry about validating anything else.

moxie0 commented on src/org/thoughtcrime/securesms/util/Rfc5724Uri.java:
```diff
@@ -0,0 +1,75 @@
+        !schemaLower.equals(SCHEMA_MMS) && !schemaLower.equals(SCHEMA_MMSTO))
+    {
+      throw new URISyntaxException(uri, "supplied URI does not contain a valid schema");
+    }
```
don't bother validating this

moxie0 commented on src/org/thoughtcrime/securesms/util/Rfc5724Uri.java:
```diff
@@ -0,0 +1,75 @@
+    return schema;
+  }
+
+  public String getRecipients() {
```
getPath

moxie0 commented on src/org/thoughtcrime/securesms/util/Rfc5724Uri.java:
```diff
@@ -0,0 +1,75 @@
+
+    if      (!uri.contains(key + "=")) return null;
+    else if (parts.length == 1)        return "";
+    else                               return URLDecoder.decode(parts[1].split("&")[0]);
```
put the query params in a map at construction time

moxie0 commented on src/org/thoughtcrime/securesms/util/Rfc5724Uri.java:
```diff
@@ -0,0 +1,75 @@
+    this.recipients = parseRecipients(uri);
+  }
+
+  private static String parseSchema(String uri) throws URISyntaxException {
```
lose the static

moxie0 commented on src/org/thoughtcrime/securesms/util/Rfc5724Uri.java:
```diff
@@ -0,0 +1,75 @@
+    return schema;
+  }
+
+  private static String parseRecipients(String uri) throws URISyntaxException {
```
lose the static

moxie0 commented on src/org/thoughtcrime/securesms/util/Rfc5724Uri.java:
```diff
@@ -0,0 +1,75 @@
+  }
+
+  private static String parseRecipients(String uri) throws URISyntaxException {
+    String[] parts = uri.split("\\?")[0].split(":");
```
limit 2

rhodey replied:
I don't understand this comment.

moxie0 replied:
split(":", 2)

moxie0 commented on src/org/thoughtcrime/securesms/util/Rfc5724Uri.java:
```diff
@@ -0,0 +1,75 @@
+  private static final String SCHEMA_SMS   = "sms";
+  private static final String SCHEMA_SMSTO = "smsto";
+  private static final String SCHEMA_MMS   = "mms";
+  private static final String SCHEMA_MMSTO = "mmsto";
```
you shouldn't need these

moxie0 commented on src/org/thoughtcrime/securesms/util/Rfc5724Uri.java:
```diff
@@ -0,0 +1,75 @@
+
+  private final String uri;
+  private final String schema;
+  private final String recipients;
```
path

moxie0 commented on src/org/thoughtcrime/securesms/util/Rfc5724Uri.java:
```diff
@@ -0,0 +1,81 @@
+  private String parseSchema() throws URISyntaxException {
+    String[] parts = uri.split(":");
+
+    if (parts.length < 1 || parts[0].length() < 1) throw new URISyntaxException(uri, "invalid schema");
```
parts.length can never be less than 1

moxie0 replied:
parts[0].isEmpty

moxie0 commented on src/org/thoughtcrime/securesms/util/Rfc5724Uri.java:
```diff
@@ -0,0 +1,81 @@
+  private String parsePath() throws URISyntaxException {
+    String[] parts = uri.split("\\?")[0].split(":", 2);
+
+    if (parts.length < 2 || parts[1].length() < 1) throw new URISyntaxException(uri, "invalid path");
```
parts[1].isEmpty

moxie0 commented on src/org/thoughtcrime/securesms/util/Rfc5724Uri.java:
```diff
@@ -0,0 +1,81 @@
+
+  private Map<String, String> parseQueryParams() throws URISyntaxException {
+    Map<String, String> queryParams = new HashMap<>();
+    if (!uri.contains("?") || uri.split("\\?").length < 2) {
```
contains should not appear in this entire class

rhodey commented on src/org/thoughtcrime/securesms/util/Rfc5724Uri.java:
```diff
@@ -0,0 +1,81 @@
+
+    for (String keyValue : uri.split("\\?")[1].split("&")) {
+      if (keyValue.contains("=")) {
+        String[] parts = keyValue.split("=");
```
`contains()` necessary for distinction between `sms:555?body, body == null` & `sms:555?body=, body == ""` because `"sms:555?body".split("=").length == 1` & `"sms:555?body=".split("=").length == 1`

moxie0 commented:
doesn't look like this has been updated from the last round of feedback

rhodey commented:
the only remaining discrepancy that I'm aware of is the use of `contains()` [here](https://github.com/WhisperSystems/TextSecure/pull/3030/files#diff-05daefa13a1a1537982fa06e7d37151dR62), `split()` alone is not enough, see:

```
String[] parts = "body=".split("=");
parts.length == 1
...
String[] parts = "body".split("=");
parts.length == 1
```

without `contains("=")`, `sms:555?body=` (legal, `body == ""`) is indistinguishable from `sms:555?body` (illegal). I can't do anything to remove the `contains()` unless we wanna be weird and say `sms:555?body=` is also illegal, then a check for `parts.length == 2` could work. split isn't all that great >:|

moxie0 commented:
it's fine without the contains

rhodey commented:
ok, then

```
sms:555?body=, body == ""
&&
sms:555?body, body == ""
&&
sms:555?a=b, body == null
```

done.

agrajaghh commented on src/org/thoughtcrime/securesms/util/Rfc5724Uri.java:
```diff
@@ -0,0 +1,80 @@
+    for (String keyValue : uri.split("\\?")[1].split("&")) {
+      String[] parts = keyValue.split("=");
+
+      if(parts.length == 1) queryParams.put(parts[0], "");
```
space :P

jpsingleton commented on src/org/thoughtcrime/securesms/util/Rfc5724Uri.java:
```diff
@@ -0,0 +1,80 @@
+  private String parsePath() throws URISyntaxException {
+    String[] parts = uri.split("\\?")[0].split(":", 2);
+
+    if (parts.length < 2 || parts[1].isEmpty()) throw new URISyntaxException(uri, "invalid path");
```
@rhodey @moxie0 I believe not having a path (phone number) is valid. Please see my comment on the related issue (https://github.com/WhisperSystems/TextSecure/issues/2578#issuecomment-120620069). Thanks!
